### PR TITLE
VMware: Use existing DVPG network in vmware_guest_network

### LINKED ISSUE #65968: vmware_guest_network: Unable to Connect the newly added interface
##### SUMMARY
Using **vmware_guest_network** I am able to add multiple network interface to the VM but unable to connect them when starting the vm. Even manually its not allowing to connect them. But when the VM is shutdown and I am selecting the 'Edit' option, its showing the interfaces are as connected.

##### ISSUE TYPE
- Bug Report

##### COMPONENT NAME
vmware_guest_network

##### ANSIBLE VERSION
```
# pip show ansible
Name: ansible
Version: 2.9.2
Summary: Radically simple IT automation
Home-page: https://ansible.com/
Author: Ansible, Inc.
Author-email: info@ansible.com
License: GPLv3+
Location: /usr/local/lib/python2.7/dist-packages
Requires: jinja2, PyYAML, cryptography

# pip show pyvmomi
Name: pyvmomi
Version: 6.7.1.2018.12
Summary: VMware vSphere Python SDK
Home-page: https://github.com/vmware/pyvmomi
Author: VMware, Inc.
Author-email: jhu@vmware.com
License: License :: OSI Approved :: Apache Software License
Location: /usr/local/lib/python2.7/dist-packages
Requires: six, requests
```

##### CONFIGURATION
```
No specific configuration required.
```

##### OS / ENVIRONMENT
Ansible Provision Server: Ubuntu 18.04 LTS
vCenter: 6.5 / 6.7
ESXi: 6.5 / 6.7

##### STEPS TO REPRODUCE
```
# cat deploy-ova.yml
---
- name: Firewall Deployment
  hosts: localhost
  connection: local
  gather_facts: false
  vars_files:
    - vsphere_creds.yml
    - vsphere_vars.yml
  tasks:
    - name: PA Firewall OVA Deployment
      vmware_deploy_ovf:
        hostname: '{{ vcenter_hostname }}'
        username: '{{ vcenter_username }}'
        password: '{{ vcenter_password }}'
        datacenter: '{{ vsphere_datacenter }}'
        cluster: '{{ vsphere_cluster }}'
        datastore: '{{ vsphere_datastore }}'
        name: '{{ vsphere_vm_name }}'
        resource_pool: '{{ vsphere_resourcepool }}'
        disk_provisioning: 'thin'
        folder: '/{{ vsphere_datacenter }}/vm/{{ vsphere_folder }}'
        validate_certs: 'no'
        networks:
          "VM Network" : "{{ vsphere_network01 }}"
          "VM Network" : "{{ vsphere_network02 }}"
        power_on: 'no'
        fail_on_spec_warnings: 'yes'
        wait_for_ip_address: 'no'
        ovf: '/root/OVA/xxxxxxxxx.ova'
      delegate_to: localhost

    - name: Attach Multiple Network Adapters
      vmware_guest_network:
        hostname: '{{ vcenter_hostname }}'
        username: '{{ vcenter_username }}'
        password: '{{ vcenter_password }}'
        datacenter: '{{ vsphere_datacenter }}'
        validate_certs: 'no'
        cluster: '{{ vsphere_cluster }}'
        name: '{{ vsphere_vm_name }}'
        networks:
          - name: "{{ vsphere_network01 }}"
            state: new
            device_type: vmxnet3
            label: "Network adapter 3"
            connected: true
            start_connected: true
          - name: "{{ vsphere_network01 }}"
            state: new
            device_type: vmxnet3
            label: "Network adapter 4"
            connected: true
            start_connected: true
          - name: "{{ vsphere_network01 }}"
            state: new
            device_type: vmxnet3
            label: "Network adapter 5"
            connected: true
            start_connected: true
          - name: "{{ vsphere_network01 }}"
            state: new
            device_type: vmxnet3
            label: "Network adapter 6"
            connected: true
            start_connected: true
          - name: "{{ vsphere_network01 }}"
            state: new
            device_type: vmxnet3
            label: "Network adapter 7"
            connected: true
            start_connected: true
          - name: "{{ vsphere_network01 }}"
            state: new
            device_type: vmxnet3
            label: "Network adapter 8"
            connected: true
            start_connected: true
          - name: "{{ vsphere_network01 }}"
            state: new
            device_type: vmxnet3
            label: "Network adapter 9"
            start_connected: true
          - name: "{{ vsphere_network01 }}"
            state: new
            device_type: vmxnet3
            label: "Network adapter 10"
            start_connected: true

    - name: Start the virtual machine
      vmware_guest:
        hostname: '{{ vcenter_hostname }}'
        username: '{{ vcenter_username }}'
        password: '{{ vcenter_password }}'
        datacenter: '{{ vsphere_datacenter }}'
        validate_certs: 'no'
        cluster: '{{ vsphere_cluster }}'
        name: '{{ vsphere_vm_name }}'
        state: poweredon
```
```
# cat vsphere_vars.yml
# vCenter Specific Variable
vsphere_datacenter: 'xxxxxxx'
vsphere_folder: 'Ansible-Demo-uskar'
vsphere_resourcepool: 'xxxxxx'
vsphere_cluster: 'xxxxxx'
vsphere_datastore: 'xxxxxx'
vsphere_vm_name: 'xxxxxx'
vsphere_network01: '192.168.0.1-management'
vsphere_network02: '192.168.1.1-deployment'
```
```
# cat vsphere_creds.yml
# vCenter information
vcenter_hostname: 'xxxxxxxxxxxx'
vcenter_username: 'ansibledemo@vsphere.local'
vcenter_password: 'Ansible@123!'
```

##### EXPECTED RESULTS
It should attach the module and connect it while starting the VM

##### ACTUAL RESULTS
Its able to connect only two interfaces mentioned in **vmware_deploy_ovf** module's function. Other interfaces are available but not connected.

### PR DESCRIPTION
**Pull request by Akasurde**
##### SUMMARY

* Handle all cases of networks

Fixes: #65968

Signed-off-by: Abhijeet Kasurde <akasurde@redhat.com>

##### ISSUE TYPE 
- Bugfix Pull Request

##### COMPONENT NAME
lib/ansible/modules/cloud/vmware/vmware_guest_network.py

### CODE REVIEW

**Comment by Akasurde** [comment]
@pgbidkar @Tomorrow9 @anusha94 Could you please review this? Thanks.

**Comment by Akasurde** [comment]
@sky-joker Could you please review this? Thanks in advance.

File: lib/ansible/modules/cloud/vmware/vmware_guest_network.py [hunk 1]
```diff
@@ -461,9 +479,31 @@ def get_network_config_spec(self, vm_obj, network_list):
                             if 'connected' in network and nic_device.connectable.connected != network['connected']:
                                 nic_device.connectable.connected = network['connected']
                                 self.change_detected = True
-                            if 'name' in network and nic_device.deviceInfo.summary != network['name']:
-                                nic_device.deviceInfo.summary = network['name']
-                                self.change_detected = True
+                            if 'name' in network:
+                                network_object = self.find_network_by_name(network_name=network['name'])[0]
+                                if network_object and hasattr(network_object, 'portKeys'):
```
Comment (sky-joker) [inline lib/ansible/modules/cloud/vmware/vmware_guest_network.py]: When changing portgroup to dvportgroup under this condition error occurs.
The cause is that the `port` key does not exist when the portgroup is assigned to the vNIC.
The following conditions are necessary.

```suggestion
                                if network_object and hasattr(network_object, 'portKeys') and hasattr(nic_spec.device.backing, 'port'):
```

File: lib/ansible/modules/cloud/vmware/vmware_guest_network.py [hunk 2]
```diff
@@ -461,9 +479,31 @@ def get_network_config_spec(self, vm_obj, network_list):
...
+                                    if network_object.config.distributedVirtualSwitch.uuid != nic_spec.device.backing.port.switchUuid:
+                                        # DistributedVirtualPortGroup
+                                        nic_spec.device.backing = vim.vm.device.VirtualEthernetCard.DistributedVirtualPortBackingInfo()
+                                        nic_spec.device.backing.port = vim.dvs.PortConnection()
+                                        nic_spec.device.backing.port.switchUuid = network_object.config.distributedVirtualSwitch.uuid
+                                        nic_spec.device.backing.port.portgroupKey = network_object.key
+                                        self.change_detected = True
+                                elif network_object and isinstance(network_object, vim.OpaqueNetwork):
```
Comment (sky-joker) [inline lib/ansible/modules/cloud/vmware/vmware_guest_network.py]: This also gives an error.
The cause is that the `opaqueNetworkId` key does not exist when the portgroup is assigned to the vNIC.
The following conditions are necessary.

```suggestion
                                elif network_object and isinstance(network_object, vim.OpaqueNetwork) and hasattr(nic_spec.device.backing, 'opaqueNetworkId'):
```

tested on vCenter/ESXi 6.7 and NSX-T 2.5

File: lib/ansible/modules/cloud/vmware/vmware_guest_network.py (hunk 1 above)
Comment (sky-joker) [inline lib/ansible/modules/cloud/vmware/vmware_guest_network.py]: When changing portgroup or NSX-T portgroup to dvportgroup under this condition error occurs.
The cause is that the `port` key does not exist when the portgroup or NSX-T portgroup is assigned to the vNIC.
The following conditions are necessary.

```suggestion
                                if network_object and hasattr(network_object, 'portKeys') and hasattr(nic_spec.device.backing, 'port'):
```

tested on vCenter/ESXi 6.7

File: lib/ansible/modules/cloud/vmware/vmware_guest_network.py (hunk 2 above)
Comment (sky-joker) [inline lib/ansible/modules/cloud/vmware/vmware_guest_network.py]: This also gives an error.
The cause is that the `opaqueNetworkId` key does not exist when the portgroup or dvportgroup is assigned to the vNIC.
The following conditions are necessary.

```suggestion
                                elif network_object and isinstance(network_object, vim.OpaqueNetwork) and hasattr(nic_spec.device.backing, 'opaqueNetworkId'):
```

tested on vCenter/ESXi 6.7 and NSX-T 2.5

**Comment by Akasurde** [comment]
@sky-joker Thanks for reviews, can you check again? Thanks.

**Comment by sky-joker** [comment]
@Akasurde 

Looks good to me :)
All tests completed successfully in my lab(integration-test included)

**Pull request merged**